sqlmesh/lsp/main.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python
+"""A Language Server Protocol (LSP) server for SQL with SQLMesh integration."""
+
+import logging
+import typing as t
+from contextlib import suppress
+from pathlib import Path
+
+from sqlmesh.core.context import Context
+from lsprotocol import types
+from pygls.server import LanguageServer
+from pygls.workspace import TextDocument
+from sqlmesh._version import __version__
+from sqlmesh.core.model import Model
+
+logger = logging.getLogger(__name__)
+
+CONTEXTS: t.Dict[str, Context] = {}
+"""A mapping of workspace paths to SQLMesh contexts."""
+
+PATHS_TO_MODELS: t.Dict[str, t.Tuple[Context, Model]] = {}
+"""A mapping of file paths to SQLMesh (context, model) tuples."""
+
+server = LanguageServer("sqlmesh_lsp", __version__)
+
+_CACHE: t.Set[str] = set()
+"""A cache of URIs for which we have already ensured a context exists."""
+
+
+def ensure_context_for_document(document: TextDocument) -> TextDocument:
+    """Ensure that a context exists for the given document if applicable by searching for a config.py or config.yml file in the parent directories."""
+    if document.uri in _CACHE:
+        return document
+    _CACHE.add(document.uri)
+    path = Path(document.path).resolve()
+    if path.suffix not in (".sql", ".py"):
+        return document
+    initial_path = path
+    while path.parents:
+        if str(path) in CONTEXTS:
+            return document
+        path = path.parent
+    path = initial_path
+    loaded = False
+    while path.parents and not loaded:
+        for ext in ("py", "yml", "yaml"):
+            config_path = path / f"config.{ext}"
+            if config_path.exists():
+                with suppress(Exception):
+                    handle = Context(paths=[f"{path}"])
+                    CONTEXTS[str(path)] = handle
+                    PATHS_TO_MODELS.update(
+                        {
+                            str(model._path.resolve()): (handle, model)
+                            for model in handle.models.values()
+                        }
+                    )
```

Comment:
Just FMI– is this pattern of defining & mutating globals common when working with LSPs?

Reply:
It's not uncomon. For me initially it's just simple and as we add complexity we can improve.